### PR TITLE
 [1/4] NVFP4 KV cache: quantization strategy abstraction and kernel

### PR DESCRIPTION
**Pull request by samuellees**
## Summary

  Part 1 of 4 for NVFP4 KV cache support on SM120 GPUs. Split from #21601 for easier review.

  **Roadmap**: PR1 (this) → PR2 (memory pool) → PR3 (attention backends) → PR4 (MTP + config + docs)

  This PR introduces the FP4 KV cache quantization strategy abstraction and quantize/dequantize utilities. Pure additions — no existing code paths are modified.

  ### Changes

  - **`fp4_kv_cache_quant_method.py`** (NEW): `FP4KVCacheQuantMethod` ABC with strategy pattern
    - `NVFP4KVMethod` — two-level scaling (per-tensor FP32 + per-block FP8 E4M3), SM100/SM120
    - `BlockFP4KVMethod` — block-wise single-level scaling (similar to MXFP4 but block_size=16)
    - `FP4_KV_CACHE_QUANT_REGISTRY` — maps recipe names to method classes
    - Methods are prefill-only (not in CUDA graph path); decode reads raw FP4 buffers via XQA kernel

  - **`kvfp4_tensor.py`** (extended):
    - `NVFP4KVQuantizeUtil` — quantize via FlashInfer `nvfp4_kv_quantize` (fallback `fp4_quantize`), dequantize via `nvfp4_kv_dequantize` (fallback PyTorch E2M1 LUT)
    - `BlockFP4KVQuantizeUtil` — block-wise FP4 with block_size=16 (renamed from `KVFP4QuantizeUtil`, backward-compat alias kept)

  - **Unit tests**: Registry, buffer shapes, quantize→dequantize roundtrip for NVFP4 (CUDA) and BlockFP4 (CPU)

  ### Design

  FP4KVCacheQuantMethod (pure compute)  →  Pool (buffer + batch dequant)  →  Backend (view adaptation)

  The quant method owns quantize/dequantize logic. The Pool (PR2) owns buffers and orchestrates batch operations. Backends (PR3) only do view/reshape.

  ### Naming Convention

  | Class | Purpose |
  |-------|---------|
  | `FP4KVCacheQuantMethod` | ABC base |
  | `NVFP4KVMethod` | Two-level scaling (global FP32 + block FP8) |
  | `BlockFP4KVMethod` | Single-level block scaling (like MXFP4, block=16) |
  | `NVFP4KVQuantizeUtil` | NVFP4 quantize/dequantize kernels |
  | `BlockFP4KVQuantizeUtil` | Block FP4 quantize/dequantize (alias: `KVFP4QuantizeUtil`) |

  ## Test plan

  - [ ] `pytest test/registered/unit/layers/quantization/test_fp4_kv_cache_quant_method.py -v`
  - [ ] Verify no existing tests broken (pure additions, backward-compat alias for `KVFP4QuantizeUtil`)

### Benchmark (Qwen3.5-35B-A3B, GSM8K, SM120)

| KV Cache        | MTP  | Accuracy |
|-----------------|------|----------|
| FP8 (fp8_e4m3)  | Yes  | 96.6%    |
| FP4 (fp4_e2m1)  | Yes  | 97.1%    |

### Throughput
|                          | FP8 KV Cache | NVFP4 KV Cache | Speedup (NVFP4 vs FP8) |
|--------------------------|--------------|----------------|------------------------|
| Prefill Latency (160K)   | 8757 ms      | 8792 ms        | 0.996x                 |
| Prefill Latency (1M)     | 142143 ms    | 142325 ms      | 0.998x                 |
| Decode Latency (1M)      | 8.4 ms       | 7.1 ms         | 1.18x                  |

### CODE REVIEW

File: python/sglang/srt/layers/quantization/kv_cache_quant_method.py [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+"""
+KV cache quantization strategy pattern.
+
+Three-player design:
+  quant_method (pure compute)  ►  Pool (buffer + batch dequant)  ►  Backend (view adaptation)
+
+Adding a new FP4 scheme only requires:
+  1. Implement a KVCacheQuantMethod subclass
+  2. Register in KV_CACHE_QUANT_REGISTRY
+  3. Add to --fp4-kv-cache-recipe choices in server_args.py
```
Comment (samuellees) [inline python/sglang/srt/layers/quantization/kv_cache_quant_method.py]: Remove

File: python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+        """Whether this method uses a per-layer global FP32 scale."""
+        return False
+
+    @abstractmethod
+    def create_buffers(
+        self, size: int, head_num: int, head_dim: int, layer_num: int, device: str
+    ) -> dict:
+        """Allocate and return a buffer dict:
+        {
+            "k_buffer": list[Tensor],       # per-layer, shape (size, head_num, head_dim//2)
```
Comment (samuellees) [inline python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py:57]: (size, head_num, head_dim//2) is specifically for FP4, if so the filename/classname should be renamed with fp4.

File: python/sglang/srt/layers/quantization/kv_cache_quant_method.py [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+        pass  # base class set_kv_buffer handles this
+
+    def dequantize_prev_kv(self, k_fp4, k_scales, v_fp4, v_scales, layer_id):
+        return k_fp4, v_fp4  # identity
+
+
+class NVFP4Method(KVCacheQuantMethod):
+    """NVFP4 two-level scaling: global FP32 + per-block FP8 E4M3.
+
+    Supported on SM100 (B200) and SM120 (RTX 5090 / RTX PRO 6000 Blackwell).
```
Comment (samuellees) [inline python/sglang/srt/layers/quantization/kv_cache_quant_method.py]: remove GPU name

File: python/sglang/srt/layers/quantization/kv_cache_quant_method.py [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+    """
+
+    name = "nvfp4"
+    scale_block_size = 16
```
Comment (samuellees) [inline python/sglang/srt/layers/quantization/kv_cache_quant_method.py]: Does this match constant name principle?

File: python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+            elif hasattr(layer, "attn"):
+                attention_layers.append(layer.attn)
+            elif hasattr(layer, "attention"):
+                if hasattr(layer.attention, "attn"):
+                    attention_layers.append(layer.attention.attn)
+
+        if not attention_layers:
+            return
+
+        # k_scales_gpu is indexed by global (absolute) layer_id.  Resize if the model
```
Comment (samuellees) [inline python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py:159]: Which scenario is this for?

File: python/sglang/srt/layers/quantization/kv_cache_quant_method.py [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+                else 1.0
+            )
+            v_scale = (
+                float(layer.v_scale)
+                if hasattr(layer, "v_scale") and layer.v_scale is not None
+                else 1.0
+            )
+            # SM100 requires a 6× adjustment to align FP4 range with hardware expectations
+            if self.sm_version == 100:
+                k_scale *= 6.0
```
Comment (samuellees) [inline python/sglang/srt/layers/quantization/kv_cache_quant_method.py]: define a constant for 6.0, for example reuse E2M1_MAX = 6.0 from kvfp4 tensor.py

File: python/sglang/srt/layers/quantization/kvfp4_tensor.py [hunk 7]
```diff
@@ -110,3 +143,91 @@ def batched_dequantize(
...
+            enable_pdl=None,
+        )
+        tensor_fp4 = tensor_fp4.view(b, m, tensor_fp4.shape[-1])
+        tensor_fp4_sf = tensor_fp4_sf.view(b, m, tensor_fp4_sf.shape[-1]).view(
+            torch.float8_e4m3fn
+        )
+        return tensor_fp4, tensor_fp4_sf, global_scale
+
+    @staticmethod
+    def dequantize(
```
Comment (samuellees) [inline python/sglang/srt/layers/quantization/kvfp4_tensor.py:221]: We should use flashinfer implementation. The API fp4_kv_dequantize

File: python/sglang/srt/layers/quantization/kvfp4_tensor.py [hunk 8]
```diff
@@ -110,3 +143,91 @@ def batched_dequantize(
...
+        global_scale_inv = 1.0 / global_scale
+        if isinstance(global_scale_inv, float):
+            global_scale_inv = torch.tensor(
+                global_scale_inv, dtype=torch.float32, device=tensor.device
+            )
+        assert (
+            global_scale_inv.device == tensor.device
+        ), "global_scale and tensor must be on the same device"
+        b, m, n = tensor.shape
+        tensor_reshaped = tensor.reshape(b * m, n)
```
Comment (samuellees) [inline python/sglang/srt/layers/quantization/kvfp4_tensor.py]: we should use fp4_kv_quantize

File: python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+        v_scale_buffer,
+        loc,
+        cache_k,
+        cache_v,
+        k_scale=None,
+        v_scale=None,
+    ) -> None:
+        pass  # base class set_kv_buffer handles this
+
+    def dequantize_prev_kv(self, k_fp4, k_scales, v_fp4, v_scales, layer_id):
```
Comment (samuellees) [inline python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py]: Should we return tensor in fp8 dtype and shape?

**Review by samuellees — COMMENTED** [review]
Does these method support cuda graph(if they need to support)?

File: test/registered/unit/layers/quantization/test_fp4_kv_cache_quant_method.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+
+        self.assertEqual(k_out.shape, (4, heads, dim))
+        self.assertEqual(k_out.dtype, torch.float8_e4m3fn)
+
+
+class TestKVFP4QuantizeUtil(CustomTestCase):
+    """Test the existing MXFP4 KVFP4QuantizeUtil roundtrip."""
+
+    def test_roundtrip_cpu(self):
+        from sglang.srt.layers.quantization.kvfp4_tensor import KVFP4QuantizeUtil
```
Comment (samuellees) [inline test/registered/unit/layers/quantization/test_fp4_kv_cache_quant_method.py]: Should we rename KVFP4QuantizeUtil to MXFP4QuantizeUtil or MXFP4QuantizeUtilBlock16?

File: python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+
+class FP4KVCacheQuantMethod(ABC):
+    """Abstract base for FP4 KV cache quantization strategies.
+
+    Owns the quantize/dequantize computation.  The Pool owns the buffers and
+    orchestrates the batch dequant loop.  Backends only do view/reshape.
+
+    Note: These methods are called during prefill (extend) only, not during
+    CUDA graph capture or decode. Decode reads raw FP4 buffers directly via
+    the XQA kernel. Therefore CUDA graph compatibility is not required here.
```
Comment (b8zhong) [inline python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py]: Question: won't this be incompatible with piecewise CUDA graph then

File: python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+        v_fp4: Tensor,
+        v_scales: Tensor,
+        layer_id: int,
+    ) -> tuple[Tensor, Tensor]:
+        """Dequantize stored FP4 KV (selected token indices already applied).
+
+        Returns:
+            (k_fp8, v_fp8): Both in torch.float8_e4m3fn dtype with shape
+            matching the input (after unpacking). These are written into the
+            shared dequant workspace buffer for the FlashInfer FP8 prefill kernel.
```
Comment (b8zhong) [inline python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py:96]: Is dequant logic (to FP8) because we don't support FP4 prefill kernel (yet)? For XQA?

File: python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py [hunk 13]
```diff
@@ -0,0 +1,421 @@
...
+        self.num_layers = num_layers
+        self.device = device
+        self.sm_version = sm_version
+        # Per-layer global FP32 scales; filled by load_scales_from_model()
+        self.k_scales_gpu = torch.ones(num_layers, dtype=torch.float32, device=device)
+        self.v_scales_gpu = torch.ones(num_layers, dtype=torch.float32, device=device)
+
+    def needs_dequant_workspace(self) -> bool:
+        return (
+            True  # prefill uses FP8 dequant workspace; future native FP4 kernel → False
```
Comment (b8zhong) [inline python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py:129]: Nevermind, just read this comment, sounds good.

File: python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+                if hasattr(layer, "k_scale") and layer.k_scale is not None
+                else 1.0
+            )
+            v_scale = (
+                float(layer.v_scale)
+                if hasattr(layer, "v_scale") and layer.v_scale is not None
+                else 1.0
+            )
+            # SM100 requires E2M1_MAX adjustment to align FP4 range with hardware expectations
+            if self.sm_version == 100:
```
Comment (b8zhong) [inline python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py:195]: QQ: can you elaborate on this, I thought the handling of the data-type itself is the same on both SM100 and SM120.

File: python/sglang/srt/layers/quantization/kvfp4_tensor.py [hunk 15]
```diff
@@ -110,3 +147,122 @@ def batched_dequantize(
         scaled = reshaped * torch.exp2(scale_exp.unsqueeze(-1))
 
         return scaled.view(b, m, n).to(dtype)
+
+
+# Backward-compatible alias (used by memory_pool.py and existing tests)
+KVFP4QuantizeUtil = BlockFP4KVQuantizeUtil
```
Comment (b8zhong) [inline python/sglang/srt/layers/quantization/kvfp4_tensor.py]: Do we need this backwards compatibility code, since I'm not sure it's not really public API

File: python/sglang/srt/layers/quantization/kvfp4_tensor.py [hunk 16]
```diff
@@ -110,3 +147,122 @@ def batched_dequantize(
...
+        try:
+            from flashinfer import nvfp4_kv_quantize
+
+            fp4_2d, scales_2d = nvfp4_kv_quantize(tensor_2d, global_scale)
+        except (ImportError, AttributeError):
+            # Fallback: fp4_quantize available on SM90+
+            from flashinfer import fp4_quantize
+
+            global_scale_inv = 1.0 / global_scale
+            fp4_2d, scales_2d = fp4_quantize(
```
Comment (b8zhong) [inline python/sglang/srt/layers/quantization/kvfp4_tensor.py:204]: If it's only use case is SM90 (since nvfp4 kv quantize expect to work on SM100/SM120), we can probably restrict the import to then (instead of try except?)

**Review by b8zhong — APPROVED** [review]

**Review by Fridge003 — APPROVED** [review]

File: python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py [hunk 17]
```diff
@@ -0,0 +1,426 @@
...
+        # Dequant workspace: shared across layers (not multiplied by num_layers), FP8
+        dq_size = head_num * head_dim * 2 * kv_size
+        return fp4_size + scale_size + dq_size
+
+
+class BlockFP4KVMethod(FP4KVCacheQuantMethod):
+    """Block-wise FP4 single-level scaling (similar to MXFP4 but block_size=16)."""
+
+    name = "blockfp4"
+    SCALE_BLOCK_SIZE = 16
```
Comment (DehuaTang) [inline python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py:312]: Hi, @samuellees ， As you know, the block size of OCP MXFP4 is 32. Why is it changed to 16 here? Does the hardware not support the block size=16 format?

Comment (samuellees) [reply python/sglang/srt/layers/quantization/fp4_kv_cache_quant_method.py:312]: Hi @DehuaTang , good question but I have no answer for that. Your question is not in the scope of this PR, please see here https://github.com/sgl-project/sglang/pull/21954/changes/BASE..942c7d6f54a56197435cea46003301a2c53fc833#diff-87c2ca9d8198ad03e3239bfef35ed25550321d88a3e97bd5820115743cd7d0a4L46
